### DISCUSSION TITLE
Allow abstract explicit interface implementation

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#7065] [Language Ideas] [answered]
Now, I can implement the following code.
```
public interface I<T>
{
    T M();
}

public abstract class C : I<int>
{
    public abstract int M();
}

public class CC : C, I<long>
{
    long I<long>.M()
    {
        ...
    }
    
    public override int M()
    {
        ...
    }
}
```
But the interface I want to explicitly implement is ```I<int>```, not ```I<long>```.

If allow abstract explicit interface implementation, I can implement the following code.
```
public interface I<T>
{
    T M();
}

public abstract class C : I<int>
{
    abstract int I<int>.M();
}

public class CC : C, I<long>
{
    public long M()
    {
        ...
    }
    
    override int I<int>.M()
    {
        ...
    }
}
```

**Comment by DavidArno** [comment]
You are looking for a change to how the language has behaved for nearly 20 years. The fact that this hasn't changed in this way likely should be setting alarm bells going in your head that you are likely disappearing down a rabbit hole of bad design.

Probably best to step back and ask yourself, "What am I aiming for here? Can I achieve this a different way?"

**Reply by ghost** [reply]
>you are likely disappearing down a rabbit hole of bad design.

Yes, I can achieve this a different way, such as the following code.
But I don't like this code, because I have to explicitly declare the method ```m()```.
```
public interface I<T>
{
    T M();
}

public abstract class C : I<int>
{
    int I<int>.M()
    {
        return m();
    }
      
   protected abstract int m();
   
}

public class CC : C, I<long>
{
    public long M()
    {
        return 1000;
    }
    
    protected override int m()
    {
        return 100;
    }
}
```

**Reply by ghost** [reply]
Another reason is that I think an abstract class is like a template, since subclasses can explicitly implement interfaces, abstract parent classes should also provide abstract explicit interface implementations.

**Comment by HaloFour** [comment, marked as answer]
Similar to this: #1618

The runtime does have support for this but C# has never exposed it.
